### PR TITLE
Tests for RRD Generation by C lib.

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

**Comment by sharady** [comment]
Code changes for tests are available in last commit.
Will perform `git-rebase` when all lib-code changes are merged.

File: librrd.h [hunk 1]
```diff
@@ -33,6 +33,10 @@
 #define RRD_FILE_ERROR          3
 #define RRD_ERROR               4
 
+#define RRD_HEADER_SIZE         31
+#define DATASOURCE_VALUE_SIZE   8
+#define META_SIZE               4
+
 typedef enum { RRD_LOCAL_DOMAIN = 0, RRD_INTER_DOMAIN } rrd_domain;
```
Comment (lindig) [inline librrd.h]: These are implementation details of the library that should not be exposed in its public interface.

File: librrd.h [hunk 2]
```diff
@@ -126,3 +130,8 @@ int             rrd_del_src(RRD_PLUGIN * plugin, RRD_SOURCE * source);
  * be called every 5 seconds by the client,
  */
 int             rrd_sample(RRD_PLUGIN * plugin);
+
+
+JSON_Value *json_for_source(RRD_SOURCE * source);
+
+JSON_Value *json_for_plugin(RRD_PLUGIN * plugin);
```
Comment (lindig) [inline librrd.h]: Again, this is information private to the implementation and must not be exposed to clients.

**Comment by lindig** [comment]
How do you know that the data produced by the library can be read by rrd-transport which is the framework used by xcp-rrd to read these files?

File: librrd.h (hunk 2 above)
Comment (sharady) [inline librrd.h]: I know, But Tests needs to use them else i have to write them again in the tests. Is there any workaround.

File: librrd.c [hunk 3]
```diff
@@ -148,7 +148,7 @@ json_for_source(RRD_SOURCE * source)
  * Generate JSON for a plugin. This is just a JSON object containing a
  * sub-objecy for every data source. 
  */
-static JSON_Value *
+JSON_Value *
```
Comment (lindig) [inline librrd.c]: You are exposing private information to the client of the library. Don't do that.

File: librrd.h [hunk 4]
```diff
@@ -33,6 +33,10 @@
 #define RRD_FILE_ERROR          3
 #define RRD_ERROR               4
 
+#define RRD_HEADER_SIZE         31
```
Comment (lindig) [inline librrd.h]: This is private information of the implementation.

File: librrd.h (hunk 2 above)
Comment (lindig) [inline librrd.h]: You can put this information in your test code but not in the public interface.

File: rrdtest.c [hunk 5]
```diff
@@ -26,87 +26,237 @@
...
 
-    v.int64 = numbers[i++ % (sizeof(numbers) / sizeof(numbers[0]))];
+RRD_SOURCE create_rrd_source(char *name, char *description, rrd_owner own, char *owner_uuid,
+        char *units, rrd_scale scl, rrd_type typ, char *min, char *max, int rrd_default, rrd_value (*f)())
+{
+    RRD_SOURCE src;
+    src.name = strdup(name);
+    src.description = strdup(description);
+    src.owner = own;
+    src.owner_uuid = strdup(owner_uuid);
```
Comment (lindig) [inline rrdtest.c]: Why do you need `strdup`?

**Comment by lindig** [comment]
While I think that internal consistency checks are useful, they are not enough to convince me that the data the library writes can be consumed by `xcp-rrdd`. The crucial bits are the binary layout of the data and the calculation of checksums. The checksums are not covered by the current tests. 

I think the best way is to write a small OCaml utility `rrdreader` that reads a file produced by `rrdclient` or `rrdtest` using the `rrd-transport` framework (that is also used by `xcp-rrdd` and reads the file). This would test the data layout and all checksums and provide the necessary integration test.

```
echo "10123" | ./rrdclient test.rrd
rrdreader test.rrd
```

I also don't like to widen the public interface of the library to expose internals for testing. I would not mind adding functions that support debugging as long as they don't reveal implementation details. If tests need to know internal knowledge, this should be kept private to the test. In any case, writing the test in OCaml would not require that.

**Comment by lindig** [comment]
If you want to pursue this, please rebase and create a clean PR.

**Comment by euanh** [comment]
Closing this PR for now as we have added some integration tests which check behaviour against rrd-transport.

**Pull request closed without merging**